File: algorithms/0376.wiggle_subsequence/wiggle_subsequence.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if (nums.size() <= 1) {
            return nums.size();
        }
        int count = 1, last = 0;
        for (int i = 0; i < nums.size() - 1; ++i) {
            const int curr = nums[i + 1] - nums[i];
            if ((curr > 0 && last <= 0)||(last >= 0 && curr < 0)) {
                ++count;
                last = curr;
            }
        }
        return count;
    }
};

File: algorithms/0376.wiggle_subsequence/wiggle_subsequence.rs
impl Solution {
    pub fn wiggle_max_length(nums: Vec<i32>) -> i32 {
        let mut last_diff = 0;
        let mut curr_diff = 0;
        let mut count = 1;
        for i in 1..nums.len() {
            curr_diff = nums[i] - nums[i - 1];
            if (curr_diff < 0 && last_diff >= 0) || (curr_diff > 0 && last_diff <= 0) {
                last_diff = curr_diff;
                count += 1;
            }
        }
        count
    }
}
